### PR TITLE
[al] ColorSet import/export: support USD displayOpacity primvar (#1015)

### PR DESCRIPTION
**Pull request by robthebloke**
UsdImaging uses displayOpacity in tandem with displayColor to set trasparency on objects. 

In AL_USDMaya, the colorset export supported either:
+ generic export of rgb/rgba data "float4[] primvars:colorSet1 = [(-1, -1, -1, -1)"
+ special case export to the "displayColor" attribute used by Hydra if the the name of the maya colorSet is "displayColor". 

This PR adds export of the "displayOpacity" attribute to the second case above, if we're exporting RGBA data

Output (manually simpliifed) now looks like this
```
    def Mesh "face"
    {   
        color3f[] primvars:displayColor = [(0, 0.8275346, 0.978), (0, 0.19235145, 0.2273255), (0, 0.8275346, 0.978), (0, 0.37345737, 0.44136077), (0, 0.17660311, 0.20871373), (0, 0.15823004, 0.187)] (
            interpolation = "uniform"
        )
        float[] primvars:displayOpacity = [1, 1, 1, 1, 1, 1] (
            interpolation = "uniform"
        )
    } 
```

### CODE REVIEW

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 1]
```diff
@@ -202,11 +208,13 @@ void MeshImportContext::gatherFaceConnectsAndVertices()
   // According to the docs for UsdGeomMesh: If 'normals' and 'primvars:normals' are both specified, the latter has precedence.
   const TfToken primvarNormalsToken("primvars:normals");
   TfToken interpolation = mesh.GetNormalsInterpolation();
+  bool isIndexed = false;
   bool hasNormalsOpinion = false;
   if(mesh.HasPrimvar(primvarNormalsToken))
   {
     UsdGeomPrimvar primvar = mesh.GetPrimvar(primvarNormalsToken);
     interpolation = primvar.GetInterpolation();
+    isIndexed = primvar.IsIndexed();
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: Nobody is using isIndexed later? 

error: variable ‘isIndexed’ set but not used [-Werror=unused-but-set-variable]

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 2]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+    fnMesh.setDisplayColors(true);
+
+    if (primvar.Get(&vtValue, m_timeCode))
+    {
+
+      //early out for primvar channels that are not Vec3/Vec4 (so exclude UVs for example)
+      if (!(vtValue.IsHolding<VtArray<GfVec3f> >() || vtValue.IsHolding<VtArray<GfVec4f> >()))
+        continue;
+
+      MStatus s;
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: Why s? Please use a more meaningful name.

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 3]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+          else if (vtValue.IsHolding<VtArray<GfVec4f> >())
+          {
+            const VtArray<GfVec4f> rawVal = vtValue.UncheckedGet<VtArray<GfVec4f> >();
+            colours.setLength(rawVal.size());
+            memcpy(&colours[0], (const float*) rawVal.cdata(), sizeof(float) * 4 * rawVal.size());
+            representation = MFnMesh::kRGBA;
+          }
+        
+          // Set colors
+          if (!fnMesh.setColors(colours, &colourSetName, representation))
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: error: ‘representation’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
`if (!fnMesh.setColors(colours, &colourSetName, representation))`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 4]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+              generateIncrementingIndices(mayaIndices, colours.length());
+            }
+            else if (interpolation == UsdGeomTokens->uniform)
+            {
+              if (colours.length() == counts.length())
+              {
+                mayaIndices.setLength(connects.length());
+                for (uint32_t i = 0, idx = 0, n = counts.length(); i < n; ++i)
+                {
+                  for (uint32_t j = 0; j < counts[i]; ++j)
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
`for (uint32_t j = 0; j < counts[i]; ++j)`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 5]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+            {
+              mayaIndices = connects;
+            }
+            else if (interpolation == UsdGeomTokens->constant)
+            {
+              mayaIndices = MIntArray(connects.length(), 0);
+            }
+          }
+
+          if (mayaIndices.length() != fnMesh.numFaceVertices())
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
`if (mayaIndices.length() != fnMesh.numFaceVertices())`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 6]
```diff
@@ -2007,7 +2109,7 @@ void MeshExportContext::copyNormalData(UsdTimeCode time, bool copyAsPrimvar)
                   missing[normalIndices[i]] = faceConnects[i];
                 }
                 else
-                if(it->second != uint32_t(faceConnects[i]))
+                if(it->second != faceConnects[i])
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
`if(it->second != faceConnects[i])`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 7]
```diff
@@ -2072,7 +2174,7 @@ void MeshExportContext::copyNormalData(UsdTimeCode time, bool copyAsPrimvar)
           bool isOrdered = true;
           for(uint32_t i = 0, n = normalIndices.length(); i < n; ++i)
           {
-            if(uint32_t(normalIndices[i]) != i)
+            if(normalIndices[i] != i)
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
`if(normalIndices[i] != i)`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp (hunk 1 above)
Comment (murphyeoin) [reply plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: This PR predates us updating our build to use the strict compilation mode - will do a pass through it... thanks

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 8]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
 
+
+void MeshImportContext::applyColourSetData()
+{
+  const std::vector<UsdGeomPrimvar> primvars = mesh.GetPrimvars();
+  for (auto it = primvars.begin(), end = primvars.end(); it != end; ++it)
+  {
+
+    const UsdGeomPrimvar& primvar = *it;
+    TfToken name, interpolation;
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]:
```
const std::vector<UsdGeomPrimvar> primvars = mesh.GetPrimvars();
for (auto it = primvars.begin(), end = primvars.end(); it != end; ++it)
{
    const UsdGeomPrimvar& primvar = *it;
}
```
Above code can be refactored out to something a bit nicer:
```
for (const UsdGeomPrimvar& primvar : mesh.GetPrimvars())
{
}
or even better with auto
for (const auto& primvar : mesh.GetPrimvars())
{
}
```

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 9]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+    SdfValueTypeName typeName;
+    int elementSize;
+
+    primvar.GetDeclarationInfo(&name, &typeName, &interpolation, &elementSize);
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: Not really sure why we creating `typeName `and `elementSize ` objects that are not used?
Can't just pass nullptr for the arguments we don't need to be populated?
`primvar.GetDeclarationInfo(&name, nullptr , &interpolation, nullptr );`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 10]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+                      colours[i] = MColor(rawValColour[i][0], rawValColour[i][1], rawValColour[i][2], rawValOpacity[i]);
+                    representation = MFnMesh::kRGBA;
+                    setCombinedDisplayAndOpacityColourSet = true;
+                  }
+                }
+              }
+            }
+            if (!setCombinedDisplayAndOpacityColourSet) 
+            {
+              const VtArray<GfVec3f> rawVal = vtValue.UncheckedGet<VtArray<GfVec3f> >();
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: why is there space between `<VtArray<GfVec3f>` and `>();`?  I've noticed this pattern has been repeated in other places in the same file. Please remove the space for all of them.

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 11]
```diff
@@ -751,8 +759,171 @@ bool MeshImportContext::applyEdgeCreases()
...
+  }
+}
+
+
 //----------------------------------------------------------------------------------------------------------------------
-void MeshImportContext::applyPrimVars(bool createUvs, bool createColours)
+void MeshImportContext::applyUVs()
 {
   const TfToken prefToken("pref");
   const std::vector<UsdGeomPrimvar> primvars = mesh.GetPrimvars();
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]:
```
const std::vector<UsdGeomPrimvar> primvars = mesh.GetPrimvars();
for (auto it = primvars.begin(), end = primvars.end(); it != end; ++it)
{
    const UsdGeomPrimvar& primvar = *it;
}
```
Above code can be refactored out to something a bit nicer:
```
for (const UsdGeomPrimvar& primvar : mesh.GetPrimvars())
{
}
or even better with auto
for (const auto& primvar : mesh.GetPrimvars())
{
}
```

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp [hunk 12]
```diff
@@ -763,17 +934,16 @@ void MeshImportContext::applyPrimVars(bool createUvs, bool createColours)
     SdfValueTypeName typeName;
     int elementSize;
     primvar.GetDeclarationInfo(&name, &typeName, &interpolation, &elementSize);
```
Comment (HamedSabri-adsk) [inline plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: Not really sure why we creating `typeName `and `elementSize ` objects that are not used?
Can't just pass nullptr for the arguments we don't need to be populated?
`primvar.GetDeclarationInfo(&name, nullptr , &interpolation, nullptr );`

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp (hunk 8 above)
Comment (murphyeoin) [reply plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: I'll make this change - but really this is just sugar IMO....in this case I was refactoring existing code and therefore adhering to the existing style.
If it's covered by our upcoming style guide, perfectly happy to adhere to it, but for now I think we should avoid too much stylistic quibbling.

File: plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp (hunk 12 above)
Comment (murphyeoin) [reply plugin/al/usdmayautils/AL/usdmaya/utils/MeshUtils.cpp]: Just ran tests, and I get a runtime error "Coding Error: in GetDeclarationInfo at line 188 of /USD-19.11/pxr/usd/lib/usdGeom/primvar.cpp -- Failed verification: ' name && typeName && interpolation && elementSize '" ... so have reverted that change

**Comment by murphyeoin** [comment]
Thanks @HamedSabri-adsk  - we will take a look at the surface normal problem. We've made some small additions/improvements to this PR internally which we were just about to push - will create a separate PR for those

**Pull request merged**